### PR TITLE
Trapping refactor

### PR DESCRIPTION
**Pull request by ascriptmaster**

### CODE REVIEW

File: test/simulator/items/shedshell.js [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+		battle.choose('p2', 'switch 2');
+		battle.commitDecisions();
+		assert.strictEqual(battle.p2.active[0].speciesid, 'heatran');
+	});
+
+	it('should allow a Pokemon to escape from moves that would trap them', function () {
+		battle = BattleEngine.Battle.construct();
+		battle.join('p1', 'Guest 1', 1, [{species: "Gengar", ability: 'levitate', moves: ['meanlook']}]);
+		battle.join('p2', 'Guest 2', 1, [
+			{species: "Trevenant", ability: 'naturalcure', item: 'shedshell', moves: ['ingrain']},
```
Comment (Marty-D) [inline test/simulator/items/shedshell.js]: I suggest using a different Pokemon here since Trevenant is naturally immune to trapping.

**Comment by Marty-D** [comment]
Looks good to me! If no one else has any comments I'll merge sometime tomorrow. :+1:

File: test/simulator/items/shedshell.js (hunk 1 above)
Comment (ascriptmaster) [inline test/simulator/items/shedshell.js]: Thanks for pointing that out. It also allowed me to notice that I straight up forgot to set the test to actually have the Pokemon move once with Mean Look / Ingrain before trying to switch

**Comment by ascriptmaster** [comment]
@Zarel can you look this over before it gets merged?

**Comment by ascriptmaster** [comment]
It's also worth mentioning that I added a Shed Shell vs Sky Drop test as per Marty's instructions.

File: battle-engine.js [hunk 2]
```diff
@@ -2960,6 +2931,39 @@ Battle = (function () {
...
+						let template = (source.illusion || source).template;
+						if (!template.abilities) continue;
+						for (let k in template.abilities) {
+							let ability = template.abilities[k];
+							if (ability === source.ability) {
+								// pokemon event was already run above so we don't need
+								// to run it again.
+								continue;
+							}
+							if ((k === 'H') && template.unreleasedHidden) {
```
Comment (TheImmortal) [inline battle-engine.js]: Unneeded parentheses

File: test/simulator/items/shedshell.js [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+const assert = require('assert');
+let battle;
+
+describe('Shed Shell', function () {
+	afterEach(function () {
+		battle.destroy();
+	});
+
+	it('should allow a Pokemon to escape a trapping abilities', function () {
```
Comment (TheImmortal) [inline test/simulator/items/shedshell.js]: >a trapping abilities

**Comment by TheImmortal** [comment]
There's inconsistency among test descs. In some cases you use "a Pokemon" and in some cases just "Pokemon". IMO just drop the "a".

**Comment by Zarel** [comment]
Looks good other than the issues TI mentioned.

**Pull request merged**